src/code/z_demo.c

```diff
@@ -971,144 +1003,145 @@ s32 Cutscene_CountNormalMasks(void) {
-                    if (cmd->textId1 != 0xFFFF) {
-                        Message_StartTextbox(play, cmd->textId1, NULL);
+                    if (cmd->altTextId1 != 0xFFFF) {
+                        Message_StartTextbox(play, cmd->altTextId1, NULL);
                     }
                 } else {
-                    Message_StartTextbox(play, cmd->base, NULL);
+                    Message_StartTextbox(play, cmd->textId, NULL);
                 }
-            } else if (cmd->type == CS_TEXTBOX_TYPE_ALL_NORMAL_MASKS) {
+            } else if (cmd->type == CS_TEXT_TYPE_ALL_NORMAL_MASKS) {
                 if (Cutscene_CountNormalMasks() == 20) {
-                    if (cmd->textId1 != 0xFFFF) {
-                        Message_StartTextbox(play, cmd->textId1, NULL);
+                    if (cmd->altTextId1 != 0xFFFF) {
+                        Message_StartTextbox(play, cmd->altTextId1, NULL);
                     }
                 } else {
-                    Message_StartTextbox(play, cmd->base, NULL);
+                    Message_StartTextbox(play, cmd->textId, NULL);
                 }
             } else {
-                Message_StartTextbox(play, cmd->base, NULL);
+                Message_StartTextbox(play, cmd->textId, NULL);
             }
-        } else {
-            goto else_label;
+            goto end;
         }
-    } else if (D_801BB128 != cmd->base) {
-        D_801BB160 = CS_TEXTBOX_TYPE_LEARN_SONG;
-        D_801BB128 = cmd->base;
-        func_80152434(play, cmd->base);
     } else {
-    else_label:
-        if (csCtx->frames >= cmd->endFrame) {
-            // The Textbox command can change the current cutscene frame, mainly to prevent advancing the cutscene when
-            // a textbox that is expected to be closed by the user is still open.
-
-            originalCsFrames = csCtx->frames;
-            talkState = Message_GetState(&play->msgCtx);
-            if ((talkState != TEXT_STATE_CLOSING) && (talkState != TEXT_STATE_NONE) && (talkState != TEXT_STATE_7) &&
-                (talkState != TEXT_STATE_8)) {
-                csCtx->frames--;
-                if ((talkState == TEXT_STATE_CHOICE) && Message_ShouldAdvance(play)) {
-                    if (play->msgCtx.choiceIndex == 0) {
-                        if (cmd->base == 0x33BD) {
-                            func_8019F230();
-                        }
+        if (sCurOcarinaAction != cmd->textId) {
+            sCutsceneTextboxType = CS_TEXT_OCARINA_ACTION;
+            sCurOcarinaAction = cmd->textId;
+            func_80152434(play, cmd->textId);
+            return;
+        }
+    }
+
+    if (csCtx->curFrame >= cmd->endFrame) {
+        // The Textbox command can change the current cutscene frame, mainly to prevent advancing the cutscene when
+        // a textbox that is expected to be closed by the user is still open.
+        endFrame = csCtx->curFrame;
```

Comment:
can you elaborate more on why this should be called endFrame?

Reply:
The code for this is a bit silly. But right at the start of this function, `(csCtx->curFrame > cmd->endFrame)` means a early return. So the only way `if (csCtx->curFrame >= cmd->endFrame) {` holds true is if `curFrame = endFrame`